### ISSUE TITLE
noqa: C901 placement for a function with a decorator changed in Python 3.8

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#311]
In GitLab by @msmolens on Oct 16, 2019, 13:20

*Please describe how you installed Flake8*

```
$ docker run -it --rm -v $(pwd):/app python:3.8 /bin/bash
root@b700655dac31:/# pip install -q flake8
root@34ce49cd2b0d:/# flake8 --version
3.7.8 (mccabe: 0.6.1, pycodestyle: 2.5.0, pyflakes: 2.1.1) CPython 3.8.0 on Linux
```

*Please provide the exact, unmodified output of `flake8 --bug-report`*

```json
{
  "dependencies": [
    {
      "dependency": "entrypoints",
      "version": "0.3"
    }
  ],
  "platform": {
    "python_implementation": "CPython",
    "python_version": "3.8.0",
    "system": "Linux"
  },
  "plugins": [
    {
      "is_local": false,
      "plugin": "mccabe",
      "version": "0.6.1"
    },
    {
      "is_local": false,
      "plugin": "pycodestyle",
      "version": "2.5.0"
    },
    {
      "is_local": false,
      "plugin": "pyflakes",
      "version": "2.1.1"
    }
  ],
  "version": "3.7.8"
}
```

*Please describe the problem or feature*

When using the `# noqa` comment to ignore `C901` errors in a decorated function, the comment previously had to be placed on the line containing the decorator:
```
@property  # noqa: C901
def data(self):
    ...
```

For the warning to be ignored when running Flake8 in **Python 3.8**, it looks like the comment must be placed on the `def` line instead:
```
@property
def data(self):  # noqa: C901
    ...
```

*If this is a bug report, please explain with examples (and example code) what you expected to happen and what actually happened.*

I'm testing using Docker with a test file `test.py` in the local directory:

```python
class MyClass(object):
    def __init__(self, name=None):
        self.name = name

    @property  # noqa: C901
    def data(self):
        if self.name is not None:
            return self.name
        return "default"
```

In the Docker container, I run the following command:
```
flake8 --max-complexity=1 /app/test.py
```

In Python 3.8, Flake8 finds the following error:
```
/app/test.py:6:5: C901 'MyClass.data' is too complex (2)
```

If I run the `python:3.7` Docker image and follow the same steps, Flake8 finds no errors.

In Python 3.8, if I move the `noqa` comment from the line with the `@property` decorator to the line with the `def data(self):` function declaration, then Flake8 finds no errors:

```python
class MyClass(object):
    def __init__(self, name=None):
        self.name = name

    @property
    def data(self):  # noqa: C901
        if self.name is not None:
            return self.name
        return "default"
```

It's unexpected that the `noqa` comment must be placed differently depending on the Python version.

**Comment by asottile** [comment]
In GitLab by @asottile on Oct 16, 2019, 14:04

yep! this was a fix in python3.8 which used to misrepresent the ast position of functions

there's nothing to do from flake8's perspective here, python3.8 is arguably the "correct" behaviour

**Comment by asottile** [comment]
In GitLab by @asottile on Oct 16, 2019, 14:04

closed

**Comment by asottile** [comment]
In GitLab by @sobolevn on Dec 19, 2019, 01:23

I have also stumbled upon this problem with several of my plugins.
I use `# noqa` comments for integration tests. Like so: https://github.com/wemake-services/wemake-python-styleguide/blob/876779d0bbbac99f36b9a21bb7e10521190540e6/tests/fixtures/noqa.py

I have decided to split (and possibly duplicate several tests) into three categories:
- universal (with correct locations on both versions)
- `<3.8` (with `noqa` as it used to be)
- `>=3.8` (with `noqa` on function and class definitions, not decorators)

I hope this comment might help other maintainers in the same situation.
